### PR TITLE
executor: fix incorrect NULL values of newly added columns in cache table

### LINKED ISSUE #42928: incorrect NULL values returned for newly added columns in cached tables
### 1. Minimal reproduce step (Required)
```sql
create table cache_add_column (f1 int);
insert into cache_add_column (f1) values (1);
alter table cache_add_column add column f2 int not null, add column f3 int default 3, add column f4 int default null;
alter table cache_add_column cache;
select * from cache_add_column;
select * from cache_add_column;
```

### 2. What did you expect to see? (Required)
```
1,0,3,NULL
```

### 3. What did you see instead (Required)
```
1,NULL,NULL,NULL
```

### 4. What is your TiDB version? (Required)
```
Release Version: v7.1.0-alpha-166-gb118a1822-dirty
Edition: Community
Git Commit Hash: b118a1822a0131d453feacb0fb1c49609cb93c40
Git Branch: master
UTC Build Time: 2023-04-11 08:27:57
GoVersion: go1.20.3
Race Enabled: false
TiKV Min Version: 6.2.0-alpha
Check Table Before Drop: true
Store: unistore
```

### PR DESCRIPTION
**Pull request by lqs**
### What problem does this PR solve?

Issue Number: close #42928

Problem Summary:

When adding new columns with default values to a table and enabling cache, querying the cached table results in the newly added columns having incorrect default values (NULL) instead of the defined default values.

### What is changed and how it works?
The PR includes a patch to fix the issue by ensuring that the correct default values are assigned to the newly added columns.

### Check List

Tests 

- [X] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Fix the issue that newly added columns in cached tables have incorrect NULL values instead of the defined default values of the column
```

### CODE REVIEW

File: executor/mem_reader.go [hunk 1]
```diff
@@ -202,7 +202,15 @@ func buildMemTableReader(ctx context.Context, us *UnionScanExec, tblReader *Tabl
...
-	rd := rowcodec.NewByteDecoder(colInfo, pkColIDs, nil, us.ctx.GetSessionVars().Location())
+
+	defVal := func(i int) ([]byte, error) {
+		d, err := table.GetColOriginDefaultValue(us.ctx, us.columns[i])
+		if err != nil {
+			return nil, err
+		}
+		return tablecodec.EncodeValue(us.ctx.GetSessionVars().StmtCtx, nil, d)
+	}
+	rd := rowcodec.NewByteDecoder(colInfo, pkColIDs, defVal, us.ctx.GetSessionVars().Location())
```
Comment (Defined2014) [inline executor/mem_reader.go:213]: Good catch, also need change code below for `indexLookUp` and `indexMerge`.
https://github.com/pingcap/tidb/blob/8b7b31005148b14b92f63b9a46cac0dcbc5c7c6f/executor/mem_reader.go#L795

And could add a UT like
```
create table cache_add_column (f1 int, index k(f1));
insert into cache_add_column (f1) values (1);
alter table cache_add_column add column f2 int not null, add column f3 int default 3, add column f4 int default null;
alter table cache_add_column cache;
select * from cache_add_column use index(k) where f1 = 1;
select * from cache_add_column use index(k) where f1 = 1;
```

Comment (lqs) [reply executor/mem_reader.go:213]: done, added a new commit

**Comment by Defined2014** [comment]
And also fix `TestIssue17767` test. (It's not an unstable test)

**Review by Defined2014 — COMMENTED** [review]
Maybe could check column has default value or not before get it.

**Comment by lqs** [comment]
just add a commit to disable strict sql mode when getting default value (as in table/tables/tables.go:SetPBColumnsDefaultValue)

**Review by Defined2014 — APPROVED** [review]

**Review by tiancaiamao — APPROVED** [review]

**Pull request merged**